File: services/userContext.js
/**
 * User Context Service
 *
 * Provides centralized user information retrieval and formatting for prompts
 * Ensures all prompts know who the user is and structure content from their perspective
 */

const { getPrimaryAccount } = require('../db/queries/accounts');
const logger = require('./logger');

/**
 * Get user context from request object
 * @param {Object} req - Express request object (should have req.user from auth middleware)
 * @returns {Promise<Object>} User context object with name, email, and formatted strings
 */
async function getUserContext(req) {
    try {
        // Get user from request (set by auth middleware)
        const user = req.user;
        
        if (!user) {
            logger.warn({ requestId: req.requestId }, 'No user found in request - user context unavailable');
            return null;
        }

        // Extract user info
        const userEmail = user.email;
        const userName = user.name || userEmail.split('@')[0];
        
        // Try to get primary account email (might be different from user email)
        let primaryAccountEmail = userEmail;
        try {
            if (user.id) {
                const primaryAccount = await getPrimaryAccount(user.id);
                if (primaryAccount && primaryAccount.account_email) {
                    primaryAccountEmail = primaryAccount.account_email;
                }
            }
        } catch (error) {
            logger.warn({ requestId: req.requestId, error: error.message }, 'Could not fetch primary account, using user email');
        }

        return {
            id: user.id,
            name: userName,
            email: userEmail,
            primaryAccountEmail: primaryAccountEmail,
            // Formatted strings for prompts
            formattedName: userName,
            formattedEmail: userEmail,
            // Context string for prompts
            contextString: `${userName} (${userEmail})`,
            // For filtering attendees
            emails: [userEmail, primaryAccountEmail].filter((e, i, arr) => arr.indexOf(e) === i) // unique
        };
    } catch (error) {
        logger.error({ requestId: req.requestId, error: error.message, stack: error.stack }, 'Error getting user context');
        return null;
    }
}

/**
 * Filter user from attendee list
 * @param {Array} attendees - Array of attendee objects
 * @param {Object} userContext - User context object from getUserContext
 * @returns {Array} Filtered attendees (excluding user)
 */
function filterUserFromAttendees(attendees, userContext) {
    if (!userContext || !attendees) {
        return attendees || [];
    }

    return attendees.filter(attendee => {
        const attendeeEmail = attendee.email || attendee.emailAddress;
        if (!attendeeEmail) return true; // Keep if no email
        
        // Filter out if email matches any of user's emails
        return !userContext.emails.some(userEmail => 
            userEmail.toLowerCase() === attendeeEmail.toLowerCase()
        );
    });
}

/**
 * Check if an email belongs to the user
 * @param {string} email - Email to check
 * @param {Object} userContext - User context object
 * @returns {boolean} True if email belongs to user
 */
function isUserEmail(email, userContext) {
    if (!userContext || !email) return false;
    
    return userContext.emails.some(userEmail => 
        userEmail.toLowerCase() === email.toLowerCase()
    );
}

/**
 * Format user context for prompts
 * @param {Object} userContext - User context object
 * @returns {string} Formatted string for prompts
 */
function formatUserContextForPrompt(userContext) {
    if (!userContext) {
        return 'the user';
    }
    
    return `${userContext.formattedName} (${userContext.formattedEmail})`;
}

/**
 * Get prompt prefix with user context
 * @param {Object} userContext - User context object
 * @param {string} assistantName - Name of the assistant (e.g., "Shadow")
 * @returns {string} Prompt prefix
 */
function getPromptPrefix(userContext, assistantName = 'Shadow') {
    if (!userContext) {
        return `You are ${assistantName}, an executive assistant preparing meeting briefs.`;
    }
    
    return `You are ${assistantName}, preparing ${userContext.formattedName} (${userContext.formattedEmail}) for their meetings.`;
}

module.exports = {
    getUserContext,
    filterUserFromAttendees,
    isUserEmail,
    formatUserContextForPrompt,
    getPromptPrefix
};

File: shadow-python/app/services/user_context.py
"""
User Context Service

Provides centralized user information retrieval and formatting for prompts
Ensures all prompts know who the user is and structure content from their perspective
"""

from typing import Optional, Dict, Any, List
from app.db.queries.accounts import get_primary_account
from app.services.logger import logger


async def get_user_context(user: Optional[Dict[str, Any]], request_id: str = None) -> Optional[Dict[str, Any]]:
    """
    Get user context from user object
    Args:
        user: User object from auth middleware
        request_id: Request ID for logging
    Returns:
        User context object with name, email, and formatted strings
    """
    try:
        if not user:
            logger.warn('No user found - user context unavailable', requestId=request_id)
            return None

        # Extract user info
        user_email = user.get('email')
        user_name = user.get('name') or user_email.split('@')[0] if user_email else 'Unknown'
        
        # Try to get all account emails (for multi-account support)
        account_emails = []
        primary_account_email = user_email
        try:
            if user.get('id'):
                from app.db.queries.accounts import get_accounts_by_user_id
                accounts = await get_accounts_by_user_id(user['id'])
                if accounts:
                    account_emails = [acc.get('account_email') for acc in accounts if acc.get('account_email')]
                    # Get primary account email
                    primary_account = await get_primary_account(user['id'])
                    if primary_account and primary_account.get('account_email'):
                        primary_account_email = primary_account['account_email']
        except Exception as error:
            logger.warn(f'Could not fetch accounts, using user email: {str(error)}', requestId=request_id)

        # Get unique emails (user email + all account emails)
        emails = list(dict.fromkeys([user_email] + account_emails)) if user_email else account_emails

        return {
            'id': user.get('id'),
            'name': user_name,
            'email': user_email,
            'primaryAccountEmail': primary_account_email,
            'formattedName': user_name,
            'formattedEmail': user_email,
            'contextString': f"{user_name} ({user_email})",
            'emails': emails
        }
    except Exception as error:
        logger.error(f'Error getting user context: {str(error)}', requestId=request_id)
        return None


def filter_user_from_attendees(attendees: List[Dict[str, Any]], user_context: Optional[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Filter user from attendee list
    Args:
        attendees: Array of attendee objects
        user_context: User context object from get_user_context
    Returns:
        Filtered attendees (excluding user)
    """
    if not user_context or not attendees:
        return attendees or []

    user_emails = [e.lower() for e in user_context.get('emails', [])]
    
    return [
        attendee for attendee in attendees
        if attendee.get('email') or attendee.get('emailAddress')
        if not any(
            user_email == (attendee.get('email') or attendee.get('emailAddress') or '').lower()
            for user_email in user_emails
        )
    ]


def is_user_email(email: str, user_context: Optional[Dict[str, Any]]) -> bool:
    """
    Check if an email belongs to the user
    Args:
        email: Email to check
        user_context: User context object
    Returns:
        True if email belongs to user
    """
    if not user_context or not email:
        return False
    
    user_emails = [e.lower() for e in user_context.get('emails', [])]
    return email.lower() in user_emails


def get_prompt_prefix(user_context: Optional[Dict[str, Any]]) -> str:
    """
    Format user context for prompts
    Args:
        user_context: User context object
    Returns:
        Formatted prefix string for prompts
    """
    if not user_context:
        return ''
    
    return f"You are preparing a brief for {user_context.get('formattedName')} ({user_context.get('formattedEmail')}). "
